File: golang/1827_min_operations/min_operations.go
/*
 * @Date: 2022-12-11
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-12-11
 * @FilePath: /algorithm/1827_min_operations/min_operations.go
 */

package main

func minOperations(nums []int) (ans int) {
	pre := nums[0] - 1
	max := func(a, b int) int {
		if b > a {
			return b
		}
		return a
	}
	for _, x := range nums {
		pre = max(pre+1, x)
		ans += pre - x
	}
	return
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		nums := []int{1, 1, 1}
		ans := 3
		assert(minOperations(nums) == ans)
	}
	{
		nums := []int{1, 5, 2, 4, 1}
		ans := 14
		assert(minOperations(nums) == ans)
	}
	{
		nums := []int{8}
		ans := 0
		assert(minOperations(nums) == ans)
	}
}

File: cpp/1598_min_operations/min_operations.cpp
/*
 * @Date: 2022-09-09
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-09-09
 * @FilePath: /algorithm/1598_min_operations/min_operations.cpp
 */

#include <assert.h>

#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  int minOperations(vector<string>& logs) {
    int depth = 0;
    for (auto& log : logs) {
      if (log == "./") {
        continue;
      } else if (log == "../") {
        if (depth > 0) {
          depth--;
        }
      } else {
        depth++;
      }
    }
    return depth;
  }
};

int main() {
  {
    vector<string> logs{"d1/", "d2/", "../", "d21/", "./"};
    int ans = 2;
    assert(Solution().minOperations(logs) == ans);
  }

  {
    vector<string> logs{"d1/", "d2/", "./", "d3/", "../", "d31/"};
    int ans = 3;
    assert(Solution().minOperations(logs) == ans);
  }

  {
    vector<string> logs{"d1/", "../", "../", "../"};
    int ans = 0;
    assert(Solution().minOperations(logs) == ans);
  }
}